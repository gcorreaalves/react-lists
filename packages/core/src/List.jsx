import React from 'react'
import PropTypes from 'prop-types'
import Core from './Core'
import ListItem from './ListItem'

const List = props => {
  const { items, itemRenderer, onItemClick } = props

  const renderItem = (item, index) => {
    const key = item.id || index

    return (
      <ListItem
        key={key}
        item={item}
        renderer={itemRenderer}
        onClick={onItemClick}
      />
    )
  }

  return (
    <Core {...props}>
      {items.map(renderItem)}
    </Core>
  )
}

List.propTypes = {
  displayLoading: PropTypes.bool,
  items: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.shape({
      data: PropTypes.arrayOf(PropTypes.object),
      header: PropTypes.string
    })
  ]).isRequired,
  itemRenderer: PropTypes.func.isRequired,
  listHeight: PropTypes.number.isRequired,
  loadingRenderer: PropTypes.func,
  reversed: PropTypes.bool,
  onBottomReached: PropTypes.func,
  onItemClick: PropTypes.func,
  onLoadMore: PropTypes.func,
  onTopReached: PropTypes.func
}

export default List
