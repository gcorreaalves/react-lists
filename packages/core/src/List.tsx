import React from 'react'
import Core, { CoreTypes } from './Core'
import ListItem from './ListItem'

interface ListTypes extends CoreTypes {
  items: Array<unknown>
  itemRenderer: React.FunctionComponent
  onItemClick?: (event: React.MouseEvent) => void
}

const List: React.FunctionComponent<ListTypes> = props => {
  const listRef = React.useRef(null)
  const { items, itemRenderer, onItemClick } = props

  const renderItem = (item, index) => {
    const key = item.id || index

    return (
      <ListItem
        key={key}
        item={item}
        renderer={itemRenderer}
        onClick={onItemClick}
      />
    )
  }

  return (
    <Core listRef={listRef} {...props}>
      {items.map(renderItem)}
    </Core>
  )
}

export default List
