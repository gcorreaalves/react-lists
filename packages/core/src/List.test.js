import React from 'react'
import { mount, render } from 'enzyme'
import List from './List'

describe('List component', () => {
  it('should render correctly', () => {
    const wrapper = mount(
      <List
        listHeight={400}
        items={['one', 'two', 'three']}
        itemRenderer={({ item }) => item}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should render 3 items', () => {
    const wrapper = render(
      <List
        listHeight={400}
        items={['one', 'two', 'three']}
        itemRenderer={({ item }) => item}
      />
    )

    expect(wrapper.find('li').length).toBe(3)
  })

  it('should render object list', () => {
    const wrapper = render(
      <List
        listHeight={400}
        items={[
          {
            id: 1,
            name: 'John Doe',
            email: 'johndoe@email.com'
          },
          {
            id: 1,
            name: 'Mary Doe',
            email: 'marydoe@email.com'
          }
        ]}
        itemRenderer={({ name, email }) => (
          <div className={'person'}>
            {name} (<i>{email}</i>)
          </div>
        )}
      />
    )

    expect(wrapper.find('.person').length).toBe(2)
  })

  it('should render an empty list', () => {
    const wrapper = render(
      <List listHeight={400} items={[]} itemRenderer={({ item }) => item} />
    )

    expect(wrapper.find('li').length).toBe(0)
  })

  it('should render reverse list', () => {
    const wrapper = mount(
      <List
        listHeight={400}
        items={['one', 'two', 'three']}
        itemRenderer={({ item }) => item}
      />
    )
    const wrapperReverse = mount(
      <List
        listHeight={400}
        items={['one', 'two', 'three']}
        itemRenderer={({ item }) => item}
        reverse
      />
    )

    expect(wrapper.find('ul').prop('style')).not.toHaveProperty(
      'flexDirection',
      'column-reverse'
    )
    expect(wrapperReverse.find('ul').prop('style')).toHaveProperty(
      'flexDirection',
      'column-reverse'
    )
  })

  it('should throw an Error if listHeader is eq or less than 0', () => {
    expect(() => {
      render(
        <List listHeight={0} items={[]} itemRenderer={({ item }) => item} />
      )
    }).toThrowError(
      new Error(
        'The attribute listHeight is required and must be greater than 0.'
      )
    )
  })

  it('should display loading component', () => {
    const wrapper = mount(
      <List
        displayLoading
        listHeight={400}
        items={['one', 'two', 'three']}
        itemRenderer={({ item }) => item}
        onLoadMore={() => {}}
      />
    )

    expect(wrapper.find('svg').length).toBe(1)
  })

  it('should display custom loading component', () => {
    const LoadingRenderer = () => <div className='loading'>loading...</div>
    const wrapper = mount(
      <List
        displayLoading
        listHeight={400}
        loadingRenderer={LoadingRenderer}
        items={['one', 'two', 'three']}
        itemRenderer={({ item }) => item}
        onLoadMore={() => {}}
      />
    )

    expect(wrapper.find(LoadingRenderer).length).toBe(1)
    expect(wrapper.find(LoadingRenderer).text()).toBe('loading...')
  })

  it('should call onItemClick', () => {
    const handleItemClick = jest.fn()
    const wrapper = mount(
      <List
        listHeight={400}
        items={['item 1', 'item 2', 'item 3']}
        itemRenderer={({ item }) => item}
        onItemClick={handleItemClick}
      />
    )

    wrapper.find('li').first().simulate('click')

    expect(handleItemClick).toHaveBeenCalledWith('item 1')
  })

  it('should call onLoadMore', () => {
    const handleLoadReached = jest.fn()
    const wrapper = mount(
      <List
        listHeight={400}
        items={[]}
        itemRenderer={({ item }) => item}
        onLoadMore={handleLoadReached}
      />
    )

    const scrollEvent = {
      target: {
        offsetHeight: 600,
        scrollHeight: 1000,
        scrollTop: 400
      }
    }

    wrapper.find('ul').first().simulate('scroll', scrollEvent)

    expect(handleLoadReached).toHaveBeenCalled()
  })

  it('should call onLoadMore on top when list is reverse', () => {
    const handleLoadReached = jest.fn()
    const wrapper = mount(
      <List
        listHeight={400}
        items={[]}
        itemRenderer={({ item }) => item}
        onLoadMore={handleLoadReached}
        reverse
      />
    )

    const scrollEvent = {
      target: {
        offsetHeight: 600,
        scrollHeight: 1000,
        scrollTop: 400
      }
    }

    wrapper.find('ul').first().simulate('scroll', scrollEvent)

    expect(handleLoadReached).toHaveBeenCalled()
  })

  it('should call onBottomReached', () => {
    const handleBottomReached = jest.fn()
    const wrapper = mount(
      <List
        listHeight={400}
        items={[]}
        itemRenderer={({ item }) => item}
        onBottomReached={handleBottomReached}
      />
    )

    const scrollEvent = {
      target: {
        offsetHeight: 600,
        scrollHeight: 1000,
        scrollTop: 400
      }
    }

    wrapper.find('ul').first().simulate('scroll', scrollEvent)

    expect(handleBottomReached).toHaveBeenCalled()
  })

  it('should call onTopReached', () => {
    const handleTopReached = jest.fn()
    const wrapper = mount(
      <List
        listHeight={400}
        items={[]}
        itemRenderer={({ item }) => item}
        onTopReached={handleTopReached}
      />
    )

    const scrollEvent = {
      target: {
        offsetHeight: 400,
        scrollHeight: 1000,
        scrollTop: 0
      }
    }

    wrapper.find('ul').first().simulate('scroll', scrollEvent)

    expect(handleTopReached).toHaveBeenCalled()
  })

  it('should call onBottomReached and onTopReached twice each', () => {
    const handleBottomReached = jest.fn()
    const handleTopReached = jest.fn()
    const wrapper = mount(
      <List
        listHeight={400}
        items={[]}
        itemRenderer={({ item }) => item}
        onBottomReached={handleBottomReached}
        onTopReached={handleTopReached}
      />
    )

    const listElement = wrapper.find('ul').first()

    const scrollEventToBottom = {
      target: {
        offsetHeight: 600,
        scrollHeight: 1000,
        scrollTop: 400
      }
    }

    const scrollEventToTop = {
      target: {
        offsetHeight: 400,
        scrollHeight: 1000,
        scrollTop: 0
      }
    }

    listElement.simulate('scroll', scrollEventToBottom)
    listElement.simulate('scroll', scrollEventToTop)
    listElement.simulate('scroll', scrollEventToBottom)
    listElement.simulate('scroll', scrollEventToTop)

    expect(handleBottomReached).toHaveBeenCalledTimes(2)
    expect(handleTopReached).toHaveBeenCalledTimes(2)
  })

  it('should call onBottomReached reverse list', () => {
    const handleBottomReached = jest.fn()
    const wrapper = mount(
      <List
        listHeight={400}
        items={[]}
        itemRenderer={({ item }) => item}
        reverse
        onBottomReached={handleBottomReached}
      />
    )

    const listElement = wrapper.find('ul').first()

    const scrollEventToBottom = {
      target: {
        offsetHeight: 600,
        scrollHeight: 1000,
        scrollTop: 0
      }
    }

    listElement.simulate('scroll', scrollEventToBottom)

    expect(handleBottomReached).toHaveBeenCalledTimes(1)
  })

  it('should call onTopReached reverse list', () => {
    const handleTopReached = jest.fn()
    const wrapper = mount(
      <List
        listHeight={400}
        items={[]}
        itemRenderer={({ item }) => item}
        reverse
        onTopReached={handleTopReached}
      />
    )

    const listElement = wrapper.find('ul').first()

    const scrollEventToTop = {
      target: {
        offsetHeight: 600,
        scrollHeight: 1000,
        scrollTop: 400
      }
    }

    listElement.simulate('scroll', scrollEventToTop)

    expect(handleTopReached).toHaveBeenCalledTimes(1)
  })

  it('should not call onBottomReached and onTopReached', () => {
    const handleBottomReached = jest.fn()
    const handleTopReached = jest.fn()
    const wrapper = mount(
      <List
        listHeight={400}
        items={[]}
        itemRenderer={({ item }) => item}
        onBottomReached={handleBottomReached}
        onTopReached={handleTopReached}
      />
    )

    const listElement = wrapper.find('ul').first()

    const scrollEvent = {
      target: {
        offsetHeight: 200,
        scrollHeight: 1000,
        scrollTop: 400
      }
    }

    listElement.simulate('scroll', scrollEvent)

    expect(handleBottomReached).not.toHaveBeenCalled()
    expect(handleTopReached).not.toHaveBeenCalled()
  })
})
