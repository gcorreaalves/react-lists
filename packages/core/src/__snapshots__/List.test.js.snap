// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should render correctly 1`] = `
<List
  itemRenderer={[Function]}
  items={
    Array [
      "one",
      "two",
      "three",
    ]
  }
  listHeight={400}
>
  <Core
    itemRenderer={[Function]}
    items={
      Array [
        "one",
        "two",
        "three",
      ]
    }
    listHeight={400}
    listRef={
      Object {
        "current": <ul
          style="list-style: none; padding: 0px; margin: 0px; width: 100%; height: 400px; overflow-y: scroll;"
        >
          <li>
            one
          </li>
          <li>
            two
          </li>
          <li>
            three
          </li>
        </ul>,
      }
    }
  >
    <div
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <ul
        onScroll={[Function]}
        style={
          Object {
            "height": 400,
            "listStyle": "none",
            "margin": 0,
            "overflowY": "scroll",
            "padding": 0,
            "width": "100%",
          }
        }
      >
        <ListItem
          item="one"
          key="0"
          renderer={[Function]}
        >
          <li
            onClick={[Function]}
          >
            <itemRenderer
              item="one"
            >
              one
            </itemRenderer>
          </li>
        </ListItem>
        <ListItem
          item="two"
          key="1"
          renderer={[Function]}
        >
          <li
            onClick={[Function]}
          >
            <itemRenderer
              item="two"
            >
              two
            </itemRenderer>
          </li>
        </ListItem>
        <ListItem
          item="three"
          key="2"
          renderer={[Function]}
        >
          <li
            onClick={[Function]}
          >
            <itemRenderer
              item="three"
            >
              three
            </itemRenderer>
          </li>
        </ListItem>
      </ul>
    </div>
  </Core>
</List>
`;
